tracing/trace.go

```diff
@@ -0,0 +1,174 @@
+		}
+		exp, err := stdouttrace.New(opts...)
+
+		if err != nil {
+			return fmt.Errorf("failed to create stdout trace exporter: %w", err)
+		}
+		tc.exporter = exp
+		return nil
+	}
+}
+
+// WithSpanExporter sets the exporter
+func WithSpanExporter(exporter sdktrace.SpanExporter) Option {
+	return func(tc *tracingConfig) error {
+		tc.exporter = exporter
+		return nil
+	}
+}
+
+// Init initializes the tracer provider. The function is meant to be called once upon
+// program setup. The reason for providing the serviceName explicitly is so that
+// this package does not depend on the observability package to be intitialized
+//
+// Best practice is to use the serviceName from the package, having previously called
+// `observability.Init`.
+func Init(opts ...Option) (ShutdownFunc, error) {
+	tracerProvider, err := NewTracerProvider(opts...)
+	if err != nil {
+		return noShutdown, fmt.Errorf("failed to create tracer provider: %w", err)
+	}
+
+	// it is imperative that this is set, since Start relies on the global tracer provider to be set.
+	// Otherwise, trace propagation will not work
+	otel.SetTracerProvider(tracerProvider)
+
+	// set global propagator to tracecontext (the default is no-op).
+	otel.SetTextMapPropagator(propagation.TraceContext{})
+
+	// Shutdown will flush any remaining spans and shut down the exporter.
+	return tracerProvider.Shutdown, nil
+}
+
+// NewTracerProvider creates a new tracer provider using the provided service name. The options
+// can and should be used to supply a span exporter. There is no default set on purpose, since
+// the choice of exporter highly depends on the environment the application is deployed in
+func NewTracerProvider(opts ...Option) (tp *sdktrace.TracerProvider, err error) {
+	// apply options
+	tracingCfg := &tracingConfig{}
+	for _, opt := range opts {
+		err = opt(tracingCfg)
+		if err != nil {
+			return nil, err
+		}
+	}
+	exporter := tracingCfg.exporter
+	if exporter == nil {
+		return nil, fmt.Errorf("no span exporter set")
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to create resource: %w", err)
+	}
+	// Register the trace exporter with a TracerProvider, using a batch
+	// span processor to aggregate spans before export.
+	bsp := sdktrace.NewBatchSpanProcessor(exporter)
+	tp = sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
+		// this will run all standard detectors to add service name and version to the resource
+		sdktrace.WithResource(resource.Default()),
+		sdktrace.WithSpanProcessor(bsp),
+	)
+	return tp, nil
+}
+
+// Start gets the tracer provider and starts a span with the provided name and options. Additional observability attributes
+// from the context are added to the span with trace.WithAttributes
+func Start(ctx context.Context, spanName string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
+	tracer := otel.GetTracerProvider().Tracer(instrumentationCodeProviderName)
```

Comment:
See my comment above about the `enabled` flag. I guess if that's false, here should be some logic to skip any action (or even before when doing the instantiation, not sure where the best place is to split paths)...

Reply:
Probably not worth the effort. Contrary to logging, tracing is/should be done at well defined paths entering parts of the stack that are any/all of

a) heavy in processing
b) not in a tight loop
c) involve downstream systems (see a)

So optimizing with enabled along the lines of the logging libraries doesn't really bring much benefit if the instrumentation isn't already at the wrong place.

Reply:
True, I was more aiming along the lines of: how does turning on / off work then? From what I saw in the accompanying PR on goProbe it seems you're always doing the call(s) right now, no matter if it's enabled or not. At _some_ point you'll have to decide if stuff is going to be triggered or not (otherwise what's the point of the flag at all), no?